exporter/prometheusremotewrite: glue up and use Write-Ahead-Log

odeke-em opened the pull request:
This change completes the WAL capability for the Prometheus Remote-Write exporter
that allows recovery of data if the prior write requests weren't yet exported.

This wires up a Write-Ahead-Log (WAL) that was implemented in PR #3597, which
was split off from PR #3017.

Note: there is very rare condition for which we can perhaps send the
same data many times and it can happen if we haven't yet truncated the
WAL, the RemoteWrite endpoint received the prior data but the process
received a Ctrl+C, kill signal. However, this will be very rare and
would have to be timed so fast and precisely.

Replaces PR #3017
Updates PR #3597
Fixes #3727
Fixes open-telemetry/wg-prometheus#9

odeke-em commented:
Kindly cc-ing @rakyll @bogdandrutu @tigrannajaryan @alolita @Aneurysm9 @dashpole. This change is split off a PR that was already approved but was deemed too heavy, please help me take a look.

tigrannajaryan commented on exporter/prometheusremotewriteexporter/README.md:
```diff
@@ -48,6 +48,10 @@ Example:
     endpoint: "https://my-cortex:7900/api/v1/push"
+    wal: # Enabling the Write-Ahead-Log for the exporter.
+        directory: ./prom_rw # The directory to store the WAL in
+        buffer_size: 100 # Option count of elements to be read from the WAL before truncating; default of 300
```
```suggestion
        buffer_size: 100 # Optional count of elements to be read from the WAL before truncating; default of 300
```

odeke-em replied:
Thank you @tigrannajaryan! I've addressed it.

tigrannajaryan commented on exporter/prometheusremotewriteexporter/exporter.go:
```diff
@@ -238,6 +278,11 @@ func (prwe *PRWExporter) export(ctx context.Context, tsMap map[string]*prompb.Ti
 			defer wg.Done()
 
 			for request := range input {
+				if err := ctx.Err(); err != nil {
```
It looks a bit unexpected that we suddenly check the context Err(). Maybe add a comment to explain why this is needed.

odeke-em replied:
I switched it up a bit, but essentially I made it so that we don't have to even attempt to compose the HTTP POST request if the context has expired, by making it instead a select statement. Please take a look.

tigrannajaryan commented on exporter/prometheusremotewriteexporter/exporter.go:
```diff
@@ -238,6 +278,11 @@ func (prwe *PRWExporter) export(ctx context.Context, tsMap map[string]*prompb.Ti
 			for request := range input {
+				if err := ctx.Err(); err != nil {
+					// TODO(@odeke-em): Perhaps log this error.
+					_ = err
```
Perhaps logger.Debug() log if unsure that it can happen frequently and flood the logs, but still important for troubleshooting?

odeke-em replied:
So that would have involved adding a logger to the system but we can actually just check that the context was cancelled, so I've done exactly that in a select. Thank you @tigrannajaryan, please take a look at the newest update.

tigrannajaryan commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -72,15 +74,22 @@ type walConfig struct {
+		return wc.BufferSize
 	}
-	return wc.NBeforeTruncation
+	return 300
```
Nit: define all config defaults as consts somewhere close to each other and close to the config struct declaration.

odeke-em replied:
Done, thanks!

tigrannajaryan reviewed (commented):
I only scanned the PR for config-related behavior. I didn't check the WAL logic.

odeke-em commented:
Thank you for the review @tigrannajaryan, I've addressed your suggestions, please take a look.

tigrannajaryan reviewed (commented):
LGTM. I would want one other reviewer to check the WAL logic before approving.

odeke-em commented:
Kindly pinging reviewers @dashpole @Aneurysm9 @rakyll @bogdandrutu @alolita. It's been almost 2 weeks since I sent this PR and almost a week since Tigran gave his approval, please help unblock me, thank you!

Aneurysm9 commented on exporter/prometheusremotewriteexporter/wal.go:
```diff
@@ -139,13 +158,137 @@ func (prwe *prweWAL) start(ctx context.Context) error {
+		atomic.AddUint64(&prwe.rWALIndex, 1)
+
+		shouldExport := false
+		select {
+		case <-timer.C:
+			shouldExport = true
+			timer.Stop()
+			timer = freshTimer()
+		default:
+			shouldExport = len(reqL) >= maxCountPerUpload
+		}
```
Does something above here block, or does this busy wait?

odeke-em replied:
This logic is so that if the timer hasn't yet ended, that we check if we need to export just based off count, given that those 2 are the conditions to export out values from the WAL.

odeke-em commented:
@Aneurysm9 I unfortunately misinformed you in our verbal chat when you asked where reads block until the WAL file is ready for reading. No, is the correct answer to your question: if the WAL has nothing in it, an ErrNotFound will be returned and the code in `(*prweWAL) readPrompbFromWAL` will spin with exponential backoff while trying to read from the WAL and then eventually return after exhausting 12 tries.

I am going to add a TODO to use fsnotify to let us know when the WAL file is ready for reading. I had it in before but there was a change we needed to add to fsnotify.

I think this change is good to go in as in, but please feel free to withdraw your approval until I have added that TODO or added fsnotify waiting, but I can do that in a follow-up PR, this change can work as is. My apologies @Aneurysm9!

odeke-em commented:
> @Aneurysm9 I unfortunately misinformed you in our verbal chat when you asked where reads block until the WAL file is ready for reading. No, is the correct answer to your question: if the WAL has nothing in it, an ErrNotFound will be returned and the code in `(*prweWAL) readPrompbFromWAL` will spin with exponential backoff while trying to read from the WAL and then eventually return after exhausting 12 tries.
> 
> I am going to add a TODO to use fsnotify to let us know when the WAL file is ready for reading. I had it in before but there was a change we needed to add to fsnotify.
> 
> I think this change is good to go in as in, but please feel free to withdraw your approval until I have added that TODO or added fsnotify waiting, but I can do that in a follow-up PR, this change can work as is. My apologies @Aneurysm9!

@Aneurysm9 I have now added file watching with fsnotify so the PR is fully complete and watches/blocks until the WAL got a write. Kindly cc-ing @tigrannajaryan to help with a final review and merge.

bogdandrutu commented:
@odeke-em sorry this needs to be moved to contrib
